### PR TITLE
ceph_argparse: put args from env before existing ones

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: https://tracker.ceph.com/issues/43795

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dillaman** [comment]
See #33213

File: src/test/CMakeLists.txt [hunk 1]
```diff
@@ -392,7 +392,7 @@ if(PERF_LOCAL_FLAGS)
   set_target_properties(ceph_perf_local PROPERTIES COMPILE_FLAGS
     ${PERF_LOCAL_FLAGS})
 endif()
-target_link_libraries(ceph_perf_local os global ${UNITTEST_LIBS})
+target_link_libraries(ceph_perf_local global ${UNITTEST_LIBS})
```
Comment (dillaman) [inline src/test/CMakeLists.txt]: Unrelated change should not be in PR.

File: src/common/ceph_argparse.cc [hunk 2]
```diff
@@ -98,15 +93,7 @@ void env_to_vec(std::vector<const char*>& args, const char *name)
...
-  bool dashdash = false;
-  std::vector<const char*> options;
-  std::vector<const char*> arguments;
-  if (split_dashdash(args, options, arguments))
-    dashdash = true;
-
-  std::vector<const char*> env_options;
-  std::vector<const char*> env_arguments;
-  std::vector<const char*> env;
+  auto [options, arguments] = split_dashdash(args);
```
Comment (dillaman) [inline src/common/ceph_argparse.cc]: We need to backport to Mimic -- was C++17 supported?

File: src/common/ceph_argparse.cc [hunk 3]
```diff
@@ -123,19 +110,21 @@ void env_to_vec(std::vector<const char*>& args, const char *name)
...
-  vector<string>::iterator i;
-  for (i = g_str_vec.begin(); i != g_str_vec.end(); ++i)
-    env.push_back(i->c_str());
-  if (split_dashdash(env, env_options, env_arguments))
-    dashdash = true;
+  std::vector<const char*> env;
+  for (const auto& s : g_str_vec) {
+    env.push_back(s.c_str());
+  }
+  auto [env_options, env_arguments] = split_dashdash(env);
```
Comment (dillaman) [inline src/common/ceph_argparse.cc]: As commented in the other PR, to me it doesn't really make sense to support `--` in `CEPH_ARGS`. There is no Ceph "global" positional argument that would apply to all CLI tools.

File: src/common/ceph_argparse.cc (hunk 2 above)
Comment (tchaikov) [reply src/common/ceph_argparse.cc]: yeah. https://github.com/ceph/ceph/blob/mimic/src/CMakeLists.txt#L149

File: src/common/ceph_argparse.cc (hunk 3 above)
Comment (tchaikov) [reply src/common/ceph_argparse.cc]: i think we need to avoid introducing a breaking change in a fix which will be backported. we can always do this in master.

File: src/test/CMakeLists.txt (hunk 1 above)
Comment (tchaikov) [reply src/test/CMakeLists.txt]: removed.

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/mimic-43795-kefu/98d20146dfb2114b640d33d2ce86db34494c1e0f/

**Comment by tchaikov** [comment]
> Honestly, even your change is not backwards-compatible by your definition since `CEPH_ARGS` is now being placed in front of CLI args. In Luminous, it seems like the core Ceph daemons used the `CEPH_ARGS` over the CLI args but librados apps usually processed the environment first then the CLI args.
> 

i am quoting Jason's comment here. as i think it's important to note that my change is not backward compatible.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by dillaman** [comment]
(passed RBD testing)

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-02-14_06:35:26-rados-wip-kefu-testing-2020-02-14-1112-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-02-15_02:54:06-rados-wip-kefu-testing-2020-02-15-0003-distro-basic-smithi/

failures are fixed by #33347 and #33348
